### PR TITLE
Add new `tinyint` and `smallint` abstract schema types to support smaller storage requirements

### PR DESCRIPTION
**Pull request by sebastienbarre**
Per preliminary discussion in #10336 this PR:
* adds new `tinyint` and `smallint` abstract schema types to support smaller storage requirements.
* ~~adds an optional `storage` field to the schema abstraction so that various flavors of integers can be supported by specific datasources. This field provides a hint as to how many bytes are required to efficiently store the contents of a column. Backend can choose to provide and honor such hint.~~

In other words, this allows for a MySQL backend to keep its `smallint(6)`, `tinyint(4)`, or `tinyint(2)` intact instead of having them turn into `int(6)`, `int(4)`, or `int(2)`. This is particularly important when a table stores a million+ rows and a column can be stored as one byte (`tinyint`) instead of 4 (`int`).

Tests are provided.

### CODE REVIEW

File: lib/Cake/Model/Datasource/DboSource.php [hunk 1]
```diff
@@ -3459,12 +3469,16 @@ public function buildColumn($column) {
 			return null;
 		}
 
-		if (!isset($this->columns[$type])) {
-			trigger_error(__d('cake_dev', 'Column type %s does not exist', $type), E_USER_WARNING);
-			return null;
+		// if a storage requirement in bytes was set, try it first (i.e. 'integer/4' before 'integer')
+		if (isset($storage) && isset($this->columns[$type . '/' . $storage])) {
+			$real = $this->columns[$type . '/' . $storage];
```
Comment (markstory) [inline lib/Cake/Model/Datasource/DboSource.php]: Why would storage lengths be embedded in the column name? I saw a 'storage' option elsewhere in these changes. A specific option seems like a better solution to me.

File: lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php [hunk 2]
```diff
@@ -389,6 +389,36 @@ public function testBuildColumn() {
...
+		);
+		$result = $this->Dbo->buildColumn($data);
+		$expected = '`testName` tinyint(4)';
+		$this->assertEquals($expected, $result);
+		$this->Dbo->columns = $restore;
+
+		$data = array(
+			'name' => 'testName',
+			'type' => 'integer',
+			'storage' => 2
```
Comment (markstory) [inline lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php]: How does this new storage option interact with `length`? Instead of tying the sizes to specific bytes, could we do something like:

```php
'my_int' => ['type' => 'integer', 'length' => 4, 'storageType' => 'small']
```

Another option could be `'storageSize' => 2` if byte are better. My concern with using specific bytes, is that we have no error handling for values like '3' or '6', and people will make mistakes that result in an incorrect schema being created. With string names, we can have a lookup of supported values and raise errors on unsupported values.

File: lib/Cake/Model/Datasource/DboSource.php (hunk 1 above)
Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/DboSource.php]: Because there is unfortunately no API in `DboSource::buildColumn($column)` that given *all* the fields in `$column` that describe the column in an abstract way (i.e. `type` *and* the new `storage` field) would return the proper real/native type like `smallint`. There *is* such method when going the other way around -- `DataSource::column()` takes a real column type and return an abstract type.

Instead, `buildColumn()` is *directly* poking at `$this->columns[$type]`, an *array* that apparently all datasources subclasses are supposed to have in place. What this means to me is that I couldn't introduce a new API like `DboSource::abstractToRealColumn()` (the opposite of `column()`) which would take the whole `$column` array of fields as a parameter instead of just `$type`, because if people have implemented their own data sources with a `$this->columns[]` array as they are supposed to, then these will no longer work, right? 

I can't introduce a function call there if there was none in the first place, there is no contract to be enforced -- datasource subclasses were never meant to call a `abstractToRealColumn()` function. So with that in mind I asked myself "how can I still use the `$this->columns[]` array and allow developers to specify a syntax there that maps an abstract type *and* a storage requirement *to* a native type?". That's why I went for a string representation that would just be the type followed by '/' followed by the storage. Does it make sense?

Alternatively I could do both. Introduce a new `abstractToRealColumn()` function that takes `$column`, not just `$type`, and if that method is not found then use the old `$this->columns` array?

File: lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php (hunk 2 above)
Comment (sebastienbarre) [inline lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php]: `length` is not an issue. Length was always confusing to MySQL users -- it has absolutely no relation to the storage requirement, it is for display purposes. CakePHP *does* make some assumptions with respect to `tinyint` that have a length of `1` to treat them as an abstract `boolean`, and I made sure this is still working (99% sure). If you had a `tinyint(1)` in your database then you would still end up with an abstract `boolean` in your schema. It's when you are using `tinyint(2)` (or `tinyint(3)`, etc, does not matter the length as long as it's not 1) that you are (still) getting an abstract `integer`. Before my change this `integer` would turn into a 4 bytes `int` in MySQL though. Now it turns back into a `tinyint` as expected, that is, one byte. 

The problem with `'storageType'='small'` is that you are still being too abstract. Who knows how many bytes you mean here, right? If somebody were to implement a new backend, they would still not know how many bytes they can safely use for that column (and not *too many* bytes).

Absolutely, I can rename it to `storageSize`.

In this implementation, if the combination of type and storage is not found, then it defaults to what was found before my change -- in that case, you get an `integer` (4 bytes). But I see what you mean for 6, because `'type' => 'integer', 'storageSize' => 6` should result in something like a `bigint`, right? My gut feeling is this is an "advanced" option, for people who know what they are doing and want tight control of storage because they have *a lot* of rows (we do) -- I would assume these people know that `bigint` exists. Maybe a `honorStorageSize` flag could be used on the datasource itself, and default to false, to bypass my code? Where would be the right place to specify such behavior?

**Comment by sebastienbarre** [comment]
@markstory thanks for the feedback I'll have a cleaner version later today.

File: lib/Cake/Model/Datasource/DboSource.php (hunk 1 above)
Comment (markstory) [inline lib/Cake/Model/Datasource/DboSource.php]: Should we add new abstract types for the various int sizes then? Right now we're doing that implicitly with a less obvious syntax than `'type => 'smallint'`. I recognize that these new types would be less 'abstract' but there is tension between abstract types and fine grained control. Adding explicit types would let us map to existing types when smallint isn't a supported type.

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/DboSource.php]: Some like byte, integer2? Honestly not sure integer3 is worth something.

Comment (markstory) [inline lib/Cake/Model/Datasource/DboSource.php]: What about `smallint` and `tinyint`?

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/DboSource.php]: OK. So no need for a `'storage'` field anymore, you would rather have the storage requirements be conveyed by 2 new abstract types, correct?

Comment (markstory) [inline lib/Cake/Model/Datasource/DboSource.php]: I think so. We can map those types to 'integer' in databases that don't support smaller ints like sqlite.

**Comment by sebastienbarre** [comment]
@markstory alright, here is a first stab at tinyint, smallint. I've tried to update the Postgres, Sqlite and Sqlserver datasource as well, but the Mysql one is the only one I can test properly. Hopefully the continuous integration will reveal issues with the other drivers. Thanks

**Comment by sebastienbarre** [comment]
Tests appear to be passing... which is surprising to me.

File: lib/Cake/Model/Datasource/Database/Mysql.php [hunk 3]
```diff
@@ -791,6 +785,12 @@ public function column($real) {
 		if (strpos($col, 'bigint') !== false || $col === 'bigint') {
 			return 'biginteger';
 		}
+		if (strpos($col, 'tinyint') !== false) {
+			return 'tinyint';
+		}
```
Comment (markstory) [inline lib/Cake/Model/Datasource/Database/Mysql.php]: :+1: this comes after the existing boolean mapping.

**Review by markstory — APPROVED** [review]
Looks good to me.

File: lib/Cake/Model/Datasource/Database/Mysql.php (hunk 3 above)
Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Mysql.php]: You bet it does :)

File: lib/Cake/Model/Datasource/Database/Sqlite.php [hunk 4]
```diff
@@ -63,13 +63,17 @@ class Sqlite extends DboSource {
...
+ *
  * @var array
  */
 	public $columns = array(
 		'primary_key' => array('name' => 'integer primary key autoincrement'),
 		'string' => array('name' => 'varchar', 'limit' => '255'),
 		'text' => array('name' => 'text'),
 		'integer' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
+		'smallint' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
+		'tinyint' => array('name' => 'integer', 'limit' => null, 'formatter' => 'intval'),
```
Comment (markstory) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: I took a look at FormHelper and I don't see an existing mappings for `integer` so adding new types shouldn't cause issues there.

Comment (chinpei215) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Shouldn't we change [$map](https://github.com/cakephp/cakephp/blob/135a24e0f1f3c842cce7a021d599a18cb77579ff/lib/Cake/View/Helper/FormHelper.php#L1238-L1245) in `FormHelper::_magicOptions`?

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: I guess I should have grepped for `'integer'`. That's a lot of hits, so I just grepped for `'decimal'` instead, assuming that would be code related to datatypes. I'm seeing the following potential candidates for update, would welcome feedback:
* `Console/Command/Task/ModelTask.php` in `fieldValidation()`,
* `Test/Fixture/DatatypeFixture.php`,
* `Test/Fixture/UnsignedFixture.php`,
* and indeed `View/Helper/FormHelper.php`.

the rest is code I touched already under `DataSource` and the corresponding tests.
What do you guys think?

Comment (markstory) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: @chinpei215 Yes, I missed that when I skimmed the code. My mistake.

Comment (chinpei215) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Probably `ModelTask` must be updated, as it have to bake numeric validation rules for new smallint/tinyint fields. Regarding the two fixtures, I think we need not to touch it, as they would be fixtures for some test cases of `decimal` fields. Instead, we may have to create `SmallintFixture` and `TinyintFixture` to add new test cases for the new field types.

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: OK I updated `ModelTask` and `FormHelper`. I'm not 100% about the rest. It looks like `DatatypeFixture` should be the right place to add `tiny_int` and `small_int` columns, since `biginteger` and `bool` are already tested there...

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: I update `DatatypeFixture`, let me know if this is satisfactory.

Comment (chinpei215) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Sorry, I didn't see the fixtures' name well. Probably, you may have to change *both* fixtures as smallint/tinyint can be declared as `usigned`.

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Thanks @chinpei215, updating `UnsignedFixture` revealed that I had forgotten to add `smallint` and `tinyint` to `$fieldParameters['unsigned']['types']` in the `Mysql` datasource.

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: We might be all set? :)

Comment (chinpei215) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: There is need to fix some related test cases using the fixtures.  `MysqlTest::buildColumnUnsignedProvider()` seems like a data provider for such a test case.

Comment (sebastienbarre) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: OK done, thanks. I have no clue who calls `buildColumnUnsignedProvider()`. There is probably some magic going on with the function name.

Comment (chinpei215) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: See `@dataProvider` annotation in the doc comment of `MysqlTest::testBuildColumnUnsigned()`. That let PHPUnit call the data provider.
https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.dataProvider

**Comment by chinpei215** [comment]
Looks good to me. Also, I confirmed it with SQL Server on my computer. Now Console/cake schema seems to work with smallint/tinyint fields of that.

**Comment by sebastienbarre** [comment]
Hmmm, the only test failing seem to be for PHP 5.4:
```
1) MemcachedEngineTest::testConfigurationConflict

Value was not read %s

Failed asserting that false matches expected 'yay'.
```
I'm a bit suspicious this is related to my change...

**Comment by chinpei215** [comment]
@sebastienbarre  I restarted the job just now. I don't think the build failure relates to your fix. We will see.

**Comment by sebastienbarre** [comment]
Thanks. This worked.

**Comment by chinpei215** [comment]
@markstory I received a related report from a user today. He said that `beke fixture` generated `biginteger` fields without default values. In my opinion, [FixtureTask::_generateRecords()](https://github.com/cakephp/cakephp/blob/9536a10d6d048a8024f5bb32548cade2c48ca956/lib/Cake/Console/Command/Task/FixtureTask.php#L338-L341) would relate to the issue, and I think smallint/tinyint types also would cause the same issue. I can fix the issue after this commit get merged, but I would like to note it here, as I think this ticket will be a good example when someone wants to add new types to 2.x branch.

**Comment by markstory** [comment]
I'll get this merged once I get a pull request for the 3.x version opened.

**Comment by sebastienbarre** [comment]
@markstory I'm a bit confused. I'm never used 3.x. Who is porting that work on 3.x?

**Comment by markstory** [comment]
@sebastienbarre I am.

File: lib/Cake/Model/Datasource/Database/Mysql.php [hunk 5]
```diff
@@ -109,6 +113,8 @@ class Mysql extends DboSource {
 /**
  * MySQL column definition
  *
+ * @link https://dev.mysql.com/doc/refman/5.7/en/data-types.html MySQL Data Types
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Mysql.php]: Move this after the ``@var``

File: lib/Cake/Model/Datasource/Database/Postgres.php [hunk 6]
```diff
@@ -51,13 +51,17 @@ class Postgres extends DboSource {
 /**
  * Columns
  *
+ * @link https://www.postgresql.org/docs/9.6/static/datatype.html PostgreSQL Data Types
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Postgres.php]: Move this after the ``@var``

File: lib/Cake/Model/Datasource/Database/Sqlite.php [hunk 7]
```diff
@@ -63,13 +63,17 @@ class Sqlite extends DboSource {
 /**
  * SQLite3 column definition
  *
+ * @link https://www.sqlite.org/datatype3.html Datatypes In SQLite Version 3
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Move this after the ``@var``

File: lib/Cake/Model/Datasource/Database/Sqlserver.php [hunk 8]
```diff
@@ -84,13 +84,17 @@ class Sqlserver extends DboSource {
 /**
  * MS SQL column definition
  *
+ * @link https://msdn.microsoft.com/en-us/library/ms187752.aspx SQL Server Data Types
```
Comment (ravage84) [inline lib/Cake/Model/Datasource/Database/Sqlserver.php]: Move this after the ``@var``

File: lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php [hunk 9]
```diff
@@ -3368,6 +3376,26 @@ public function buildColumnUnsignedProvider() {
 					'default' => 1
 				),
 				'`testName` decimal UNSIGNED DEFAULT 1'
+			),
+			//set #9
```
Comment (ravage84) [inline lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php]: Instead of comments, you can use the array keys for documenting the data set.
The advantage is that it would show the name of the data set in case of an error in the error message.
It might even show it in the coverage, not sure though.

**Review by ravage84 — COMMENTED** [review]
Minor nitpicking from me, as usual 😼

File: lib/Cake/Model/Datasource/Database/Sqlite.php (hunk 7 above)
Comment (dereuromark) [inline lib/Cake/Model/Datasource/Database/Sqlite.php]: Should be done by a sniffer later. Now Just unnecessary Work and waste of time IMO.

File: lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php (hunk 9 above)
Comment (chinpei215) [inline lib/Cake/Test/Case/Model/Datasource/Database/MysqlTest.php]: But it has also some disadvantage. Even if people uses the same key, PHP doesn't complain it. Some important data might be not provided in such cases. If our code sniffer cannot complain it, I am not in favor of that.

**Comment by sebastienbarre** [comment]
@ravage84 I moved the @link. As far as "//set #9" is concerned, it's a bit above my head, so I'll let you guys take care of it if needed?
